设置文件版本信息

Howard20181 opened the pull request:
不固定为 0.0.0.0，可以不打开应用时读取版本信息
TouchHelper 版本检查和 Updater 检查更新时可以直接从文件读取版本号

---

其他更改：
1. 用 MP_VERSION_STRING 取代 MP_VERSION_TAG，前者不包含开头的 v 字符。现在始终定义 MP_*_VERSION，通过 MP_VERSION_STRING 区分开发版本和发布版本。
2. publish.py 改为接受版本字符串作为参数，release.yml 负责从标签名中提取版本字符串。publish.py 将版本号相关参数原样传给 msbuild 做进一步处理。
3. 如果不指定版本字符串但指定了版本号，将根据版本号自动生成版本字符串。

Howard20181 committed:
设置文件版本信息

不固定为 0.0.0.0，可以不打开应用时读取版本信息
TouchHelper 版本检查和 Updater 检查更新时可以直接从文件读取版本号

Howard20181 commented:
对 VersionTag 的理解有误，我以为是 alpha dev canary 这种版本类型标记，实际上是 v0.12.1 这样的完整版本号，这样的话不应该合并到版本号上

Blinue commented:
谢谢贡献！目前的方案是发布脚本中临时替换 rc 中的版本号，所以发布的版本是有文件版本号的。在 rc 中用宏定义应该是更好的解决方案，但我不熟悉 rc 所以没采用。

https://github.com/Blinue/Magpie/blob/dc6c65fe14cb2828623f2f1e42d89e5026310c0f/scripts/publish.py#L64-L92

> TouchHelper 版本检查和 Updater 检查更新时可以直接从文件读取版本号

没必要，版本号已经保存成常量。

https://github.com/Blinue/Magpie/blob/dc6c65fe14cb2828623f2f1e42d89e5026310c0f/src/Magpie/UpdateService.cpp#L28-L34

Howard20181 committed:
设置文件版本号时不考虑MP_VERSION_TAG

同时从 public.py 中移除替换逻辑

Howard20181 commented:
我发现 Windows 鼠标停留预览的版本号是读取 `FILEVERSION` 的
属性窗口中文件版本号是读取 `FILEVERSION`，产品版本号读取的是 `StringFileInfo` 里的 `ProductVersion`
实际上 `StringFileInfo` 里的 `FileVersion` 是没用的
不过文档中这4个都要求存在，但不存在也不会出错
https://learn.microsoft.com/en-us/windows/win32/menurc/versioninfo-resource
https://learn.microsoft.com/en-us/windows/win32/menurc/stringfileinfo-block

Blinue committed:
chore: RC 优化
1. 用 MP_VERSION_STRING 取代 MP_VERSION_TAG，前者不包含开头的 v 字符
2. RC 文件不再支持 VS 编辑，删除了冗余代码。区域由非特定语言改为 en-US
3. STRINGIFY 和 WIDEN_STRINGIFY 宏移到通用头文件

Blinue commented:
RC 中用宏定义的缺点是不能再用 Visual Studio 编辑 RC 了，不过我们没有复杂的资源所以可以接受。

Blinue committed:
chore: 简化 resource.h

Blinue committed:
chore: 删除 APSTUDIO_READONLY_SYMBOLS 宏
这个宏供 VS 资源编辑器使用

Blinue committed:
chore: 始终定义 MP_MEOW_VERSION
通过 MP_VERSION_STRING 区分开发版本和发布版本

Blinue committed:
chore: 简化版本字符串提取
使用了正则表达式的正向先行断言

Blinue commented:
@all-contributors
Please add @Howard20181 for code.
